File: scripts/fix-cors-wildcards.js
#!/usr/bin/env node
/**
 * Automated CORS Wildcard Security Fix
 * Replaces insecure wildcard CORS with secure handleCORS()
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

// Configuration
const API_DIR = path.join(__dirname, '../api');
const BACKUP_DIR = path.join(__dirname, '../.cors-backups-' + Date.now());
const MIDDLEWARE_PATH = '../middleware/cors-handler';

// Statistics
const stats = {
  fixed: 0,
  skipped: 0,
  errors: 0,
  files: []
};

// Create backup directory
if (!fs.existsSync(BACKUP_DIR)) {
  fs.mkdirSync(BACKUP_DIR, { recursive: true });
}

console.log('🔧 CORS Wildcard Security Fix');
console.log('====================================\n');
console.log(`📦 Backups: ${BACKUP_DIR}\n`);

// Find all files with wildcard CORS
function findFilesWithWildcardCORS(dir) {
  const files = [];

  function walk(directory) {
    const items = fs.readdirSync(directory);

    for (const item of items) {
      const fullPath = path.join(directory, item);
      const stat = fs.statSync(fullPath);

      if (stat.isDirectory()) {
        walk(fullPath);
      } else if (item.endsWith('.js') || item.endsWith('.ts')) {
        const content = fs.readFileSync(fullPath, 'utf8');
        if (content.includes("'Access-Control-Allow-Origin'") && content.includes("'*'")) {
          files.push(fullPath);
        }
      }
    }
  }

  walk(dir);
  return files;
}

// Calculate relative path for require()
function getRelativePath(from, to) {
  const relative = path.relative(path.dirname(from), to).replace(/\\/g, '/');
  return relative.startsWith('.') ? relative : `./${relative}`;
}

// Fix a single file
function fixFile(filePath) {
  console.log(`Processing: ${path.relative(API_DIR, filePath)}`);

  try {
    let content = fs.readFileSync(filePath, 'utf8');

    // Skip if already uses handleCORS
    if (content.includes('handleCORS')) {
      console.log('  ⊙ Already uses secure CORS - SKIPPED\n');
      stats.skipped++;
      return;
    }

    // Create backup
    const backupPath = path.join(BACKUP_DIR, path.basename(filePath) + '.bak');
    fs.writeFileSync(backupPath, content);

    // Calculate correct import path
    const middlewarePath = path.join(__dirname, MIDDLEWARE_PATH);
    const relativePath = getRelativePath(filePath, middlewarePath);

    // Add import if not present
    if (!content.includes('cors-handler')) {
      // Find where to insert (after last require)
      const requireRegex = /const .+ = require\(.+\);?\n/g;
      const matches = [...content.matchAll(requireRegex)];

      if (matches.length > 0) {
        const lastMatch = matches[matches.length - 1];
        const insertPos = lastMatch.index + lastMatch[0].length;
        const importStatement = `const { handleCORS } = require('${relativePath}');\n`;

        content = content.slice(0, insertPos) + importStatement + content.slice(insertPos);
      }
    }

    // Remove old CORS headers and replace with handleCORS
    const oldCorsPattern = /  \/\/ (Set )?CORS headers\n  res\.setHeader\('Access-Control-Allow-Origin',.*?\n  res\.setHeader\('Access-Control-Allow-Methods',.*?\n  res\.setHeader\('Access-Control-Allow-Headers',.*?\n(  res\.setHeader\('Access-Control-Allow-Credentials',.*?\n)?\n  (\/\/ Handle OPTIONS request\n  )?if \(req\.method === 'OPTIONS'\) \{\n    return res\.status\((200|204)\)\.end\(\);\n  \}/gs;

    if (oldCorsPattern.test(content)) {
      content = content.replace(oldCorsPattern, '  // Apply secure CORS\n  if (handleCORS(req, res)) return;');
    } else {
      // Try simpler pattern
      const simplePattern = /  res\.setHeader\('Access-Control-Allow-Origin', '\*'\);[\s\S]*?if \(req\.method === 'OPTIONS'\) \{\n    return res\.status\(\d+\)\.end\(\);\n  \}/;

      if (simplePattern.test(content)) {
        content = content.replace(simplePattern, '  // Apply secure CORS\n  if (handleCORS(req, res)) return;');
      } else {
        console.log('  ⚠ Pattern not matched - SKIPPED\n');
        stats.skipped++;
        return;
      }
    }

    // Write fixed content
    fs.writeFileSync(filePath, content);

    console.log('  ✓ FIXED\n');
    stats.fixed++;
    stats.files.push(path.relative(API_DIR, filePath));

  } catch (error) {
    console.log(`  ✗ ERROR: ${error.message}\n`);
    stats.errors++;
  }
}

// Main execution
try {
  const files = findFilesWithWildcardCORS(API_DIR);

  if (files.length === 0) {
    console.log('✓ No CORS wildcards found! System is secure.\n');
    process.exit(0);
  }

  console.log(`Found ${files.length} files with wildcard CORS\n`);

  // Process each file
  for (const file of files) {
    fixFile(file);
  }

  // Summary
  console.log('====================================');
  console.log('📊 SUMMARY');
  console.log('====================================');
  console.log(`✓ Fixed:   ${stats.fixed} files`);
  console.log(`⊙ Skipped: ${stats.skipped} files`);
  console.log(`✗ Errors:  ${stats.errors} files`);
  console.log(`\n📦 Backups: ${BACKUP_DIR}\n`);

  if (stats.fixed > 0) {
    console.log('✓ CORS security fixes applied successfully!\n');
    console.log('Next steps:');
    console.log('1. Review changes: git diff');
    console.log('2. Test endpoints: npm test');
    console.log('3. Commit: git add . && git commit -m "security: Fix CORS wildcard vulnerabilities"');
    console.log('4. Deploy to production\n');

    // Show fixed files
    if (stats.files.length <= 10) {
      console.log('Fixed files:');
      stats.files.forEach(f => console.log(`  - ${f}`));
    } else {
      console.log(`Fixed ${stats.files.length} files (use git diff to see all changes)`);
    }
  }

} catch (error) {
  console.error('Fatal error:', error);
  process.exit(1);
}

File: scripts/fix-cors-wildcards.py
#!/usr/bin/env python3
"""
Batch fix CORS wildcard (*) in all API files.
Replaces res.setHeader('Access-Control-Allow-Origin', '*') with getCorsOrigin(req).
Adds import of cors.js module at top of each file.
"""

import os
import re
import sys

API_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'api')
MIDDLEWARE_DIR = os.path.join(API_DIR, '_middleware')

# Patterns to replace
PATTERNS = [
    # Pattern 1: res.setHeader('Access-Control-Allow-Origin', '*')
    (
        re.compile(r"res\.setHeader\(\s*['\"]Access-Control-Allow-Origin['\"]\s*,\s*['\"]?\*['\"]?\s*\)"),
        "res.setHeader('Access-Control-Allow-Origin', getCorsOrigin(req))"
    ),
    # Pattern 2: res.setHeader('Access-Control-Allow-Origin', req.headers.origin || '*')
    (
        re.compile(r"res\.setHeader\(\s*['\"]Access-Control-Allow-Origin['\"]\s*,\s*req\.headers\.origin\s*\|\|\s*['\"]?\*['\"]?\s*\)"),
        "res.setHeader('Access-Control-Allow-Origin', getCorsOrigin(req))"
    ),
]

# Pattern for object literal style: 'Access-Control-Allow-Origin': '*'
OBJ_PATTERN = re.compile(r"(['\"])Access-Control-Allow-Origin\1\s*:\s*['\"]?\*['\"]?")
OBJ_REPLACEMENT = "'Access-Control-Allow-Origin': getCorsOrigin(req)"

# Import line to add
IMPORT_TEMPLATE = "const {{ getCorsOrigin }} = require('{}');\n"

# Already has import
IMPORT_CHECK = re.compile(r"require\(['\"].*?cors['\"]?\)")

def get_relative_path(filepath):
    """Calculate relative path from file to api/_middleware/cors"""
    file_dir = os.path.dirname(filepath)
    rel = os.path.relpath(MIDDLEWARE_DIR, file_dir)
    return rel.replace(os.sep, '/') + '/cors'

def process_file(filepath):
    """Process a single JS file for CORS fixes."""
    with open(filepath, 'r', encoding='utf-8', errors='replace') as f:
        content = f.read()

    original = content
    changes = 0

    # Apply pattern replacements
    for pattern, replacement in PATTERNS:
        content, n = pattern.subn(replacement, content)
        changes += n

    # Apply object literal pattern
    if OBJ_PATTERN.search(content):
        content, n = OBJ_PATTERN.subn(OBJ_REPLACEMENT, content)
        changes += n

    if changes == 0:
        return 0, "No CORS wildcards found"

    # Add import if not already present
    if not IMPORT_CHECK.search(content) and 'getCorsOrigin' not in original:
        rel_path = get_relative_path(filepath)
        import_line = IMPORT_TEMPLATE.format(rel_path)

        # Find good insertion point (after last require/import at top)
        lines = content.split('\n')
        insert_idx = 0
        for i, line in enumerate(lines):
            stripped = line.strip()
            if stripped.startswith('require(') or stripped.startswith("const ") and 'require(' in stripped:
                insert_idx = i + 1
            elif stripped.startswith('//') or stripped.startswith('/*') or stripped.startswith('*') or stripped == '' or stripped.startswith("'use strict"):
                if insert_idx == 0:
                    continue
            elif insert_idx > 0:
                break

        # If no requires found, insert after any initial comments/blank lines
        if insert_idx == 0:
            for i, line in enumerate(lines):
                stripped = line.strip()
                if stripped and not stripped.startswith('//') and not stripped.startswith('/*') and not stripped.startswith('*') and stripped != '':
                    insert_idx = i
                    break

        lines.insert(insert_idx, import_line.rstrip())
        content = '\n'.join(lines)

    with open(filepath, 'w', encoding='utf-8') as f:
        f.write(content)

    return changes, f"{changes} replacement(s)"

def main():
    if not os.path.isdir(API_DIR):
        print(f"ERROR: {API_DIR} not found")
        sys.exit(1)

    total_files = 0
    total_changes = 0
    fixed_files = 0

    # Walk all JS files in api/
    for root, dirs, files in os.walk(API_DIR):
        # Skip _middleware directory itself
        if '_middleware' in root.split(os.sep):
            continue

        for filename in sorted(files):
            if not filename.endswith('.js'):
                continue

            filepath = os.path.join(root, filename)
            total_files += 1

            try:
                changes, status = process_file(filepath)
                if changes > 0:
                    fixed_files += 1
                    total_changes += changes
                    rel = os.path.relpath(filepath, os.path.dirname(API_DIR))
                    print(f"  FIXED: {rel} - {status}")
            except Exception as e:
                rel = os.path.relpath(filepath, os.path.dirname(API_DIR))
                print(f"  ERROR: {rel} - {e}")

    print(f"\nSummary: {fixed_files} files fixed, {total_changes} total replacements, {total_files} files scanned")

if __name__ == '__main__':
    main()
